File: Data Structures & Algorithms/generate-parentheses/submission-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:

        def helper(lp, rp, path):
            result = list()
    
            if lp == 0 and rp == 0:
                return [path]

            if lp > 0:
                result += helper(lp - 1, rp, path + "(")
            if lp < rp:
                result += helper(lp, rp-1, path + ")")
            
            return result

        return helper(n, n, "")



            

File: Data Structures & Algorithms/reverse-bits/submission-1.cpp
class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t result = 0;

        for (int i = 0; i < 32; i++) {

            uint32_t sample = ( n & (1u << i) );
            int amountToMove = (31 - i) - i;
            if (amountToMove < 0){
                result |= sample >> std::abs(amountToMove);
            }
            else {
                result |= sample << std::abs(amountToMove);
            }

        }

        return result;


        // 1 0 1 0 1 0
        // 5   3 2 

    }
};
